### ISSUE TITLE
Functions on pair "can't unify"

### ISSUE THREAD

**Issue by xsebek** [byorgey/swarm#130]
**Describe the bug**
Pair functions `fst` and `snd` seem to have wrong signature `u0 * (u1 -> u0)`.

**To Reproduce**
```haskell
> fst
fst : ∀ a0 a1. a0 * (a1 -> a0)
> fst (1,2)
┌──────────────────Error─────────────────┐
│Can't unify u0 * (u1 -> u0) and u2 -> u3│
└────────────────────────────────────────┘
```
It looks like the `fst` is in fact something like
```haskell
> (0, \c. 1)
(0, \c. 1) : ∀ a0. int * (a0 -> int)
```
But given those functions are the only way to extract pairs, I can't see if they are actually a pair :smile: 

**Expected behavior**
Obviously, the signature should be `fst : a*b->a` and that should unify.

**Comment by byorgey** [comment]
Whoops!  I guess I never actually tested these at all.  My guess is it's a precedence error in the type parser... yep, I accidentally made function arrow have *higher* precedence than product, but it should be the other way around.

**Comment by byorgey** [comment]
Oh, I know what happened: I surely tried out `fst` and `snd` when I first added them, and they worked fine.  But when I added them, the `inferConst` function was just returning some manually constructed type ASTs, so of course it had the correct type.  Later, I added a quasiquoter for types, and switched the `inferConst` function to use it, at which point the types became wrong.

And this is exactly why we need to add more tests...
